### PR TITLE
`data_tabulate()` gains `by` argument for crosstables

### PR DESCRIPTION
**Pull request by strengejacke**
And `include_na` to include/omit missings.

``` r
library(datawizard)
data(efc)

set.seed(123)
efc$weights <- abs(rnorm(n = nrow(efc), mean = 1, sd = 0.5))
efc$e16sex[sample.int(nrow(efc), 5)] <- NA

data_tabulate(efc$c172code)
#> carer's level of education (efc$c172code) <numeric>
#> # total N=100 valid N=90
#> 
#> Value |  N | Raw % | Valid % | Cumulative %
#> ------+----+-------+---------+-------------
#> 1     |  8 |  8.00 |    8.89 |         8.89
#> 2     | 66 | 66.00 |   73.33 |        82.22
#> 3     | 16 | 16.00 |   17.78 |       100.00
#> <NA>  | 10 | 10.00 |    <NA> |         <NA>

data_tabulate(efc$c172code, include_na = FALSE)
#> carer's level of education (efc$c172code) <numeric>
#> # total N=90 valid N=90
#> 
#> Value |  N | Raw % | Valid % | Cumulative %
#> ------+----+-------+---------+-------------
#> 1     |  8 |  8.89 |    8.89 |         8.89
#> 2     | 66 | 73.33 |   73.33 |        82.22
#> 3     | 16 | 17.78 |   17.78 |       100.00

data_tabulate(efc$c172code, weights = efc$weights)
#> carer's level of education (efc$c172code) <numeric>
#> # total N=105 valid N=92 (weighted)
#> 
#> Value |  N | Raw % | Valid % | Cumulative %
#> ------+----+-------+---------+-------------
#> 1     | 10 |  9.52 |   10.87 |        10.87
#> 2     | 67 | 63.81 |   72.83 |        83.70
#> 3     | 15 | 14.29 |   16.30 |       100.00
#> <NA>  | 13 | 12.38 |    <NA> |         <NA>

data_tabulate(efc$c172code, include_na = FALSE, weights = efc$weights)
#> carer's level of education (efc$c172code) <numeric>
#> # total N=92 valid N=92 (weighted)
#> 
#> Value |  N | Raw % | Valid % | Cumulative %
#> ------+----+-------+---------+-------------
#> 1     | 10 | 10.87 |   10.87 |        10.87
#> 2     | 67 | 72.83 |   72.83 |        83.70
#> 3     | 15 | 16.30 |   16.30 |       100.00


data_tabulate(efc$c172code, by = efc$e16sex, proportions = "cell")
#> efc$c172code |       male |     female |     <NA> | Total
#> -------------+------------+------------+----------+------
#> 1            |  5  (5.0%) |  2  (2.0%) | 1 (1.0%) |     8
#> 2            | 31 (31.0%) | 33 (33.0%) | 2 (2.0%) |    66
#> 3            |  4  (4.0%) | 11 (11.0%) | 1 (1.0%) |    16
#> <NA>         |  5  (5.0%) |  4  (4.0%) | 1 (1.0%) |    10
#> -------------+------------+------------+----------+------
#> Total        |         45 |         50 |        5 |   100

data_tabulate(efc$c172code, by = efc$e16sex, proportions = "cell", include_na = FALSE)
#> efc$c172code |       male |     female | Total
#> -------------+------------+------------+------
#> 1            |  5  (5.8%) |  2  (2.3%) |     7
#> 2            | 31 (36.0%) | 33 (38.4%) |    64
#> 3            |  4  (4.7%) | 11 (12.8%) |    15
#> -------------+------------+------------+------
#> Total        |         40 |         46 |    86

data_tabulate(efc$c172code, by = efc$e16sex, proportions = "cell", weights = efc$weights)
#> efc$c172code |       male |     female |     <NA> | Total
#> -------------+------------+------------+----------+------
#> 1            |  5  (4.8%) |  3  (2.9%) | 2 (1.9%) |    10
#> 2            | 32 (30.5%) | 32 (30.5%) | 3 (2.9%) |    67
#> 3            |  3  (2.9%) | 11 (10.5%) | 1 (1.0%) |    15
#> <NA>         |  8  (7.6%) |  5  (4.8%) | 1 (1.0%) |    14
#> -------------+------------+------------+----------+------
#> Total        |         48 |         51 |        7 |   105

data_tabulate(efc$c172code, by = efc$e16sex, proportions = "cell", include_na = FALSE, weights = efc$weights)
#> efc$c172code |       male |     female | Total
#> -------------+------------+------------+------
#> 1            |  5  (5.8%) |  3  (3.5%) |     8
#> 2            | 32 (37.2%) | 32 (37.2%) |    64
#> 3            |  3  (3.5%) | 11 (12.8%) |    14
#> -------------+------------+------------+------
#> Total        |         40 |         46 |    86


data_tabulate(efc, "c172code", by = efc$e16sex, proportions = "row")
#> c172code |       male |     female |      <NA> | Total
#> ---------+------------+------------+-----------+------
#> 1        |  5 (62.5%) |  2 (25.0%) | 1 (12.5%) |     8
#> 2        | 31 (47.0%) | 33 (50.0%) | 2  (3.0%) |    66
#> 3        |  4 (25.0%) | 11 (68.8%) | 1  (6.2%) |    16
#> <NA>     |  5 (50.0%) |  4 (40.0%) | 1 (10.0%) |    10
#> ---------+------------+------------+-----------+------
#> Total    |         45 |         50 |         5 |   100

data_tabulate(efc, "c172code", by = efc$e16sex, proportions = "row", include_na = FALSE)
#> c172code |       male |     female | Total
#> ---------+------------+------------+------
#> 1        |  5 (71.4%) |  2 (28.6%) |     7
#> 2        | 31 (48.4%) | 33 (51.6%) |    64
#> 3        |  4 (26.7%) | 11 (73.3%) |    15
#> ---------+------------+------------+------
#> Total    |         40 |         46 |    86

data_tabulate(efc, "c172code", by = efc$e16sex, proportions = "row", weights = efc$weights)
#> c172code |       male |     female |      <NA> | Total
#> ---------+------------+------------+-----------+------
#> 1        |  5 (50.0%) |  3 (30.0%) | 2 (20.0%) |    10
#> 2        | 32 (47.8%) | 32 (47.8%) | 3  (4.5%) |    67
#> 3        |  3 (20.0%) | 11 (73.3%) | 1  (6.7%) |    15
#> <NA>     |  8 (57.1%) |  5 (35.7%) | 1  (7.1%) |    14
#> ---------+------------+------------+-----------+------
#> Total    |         48 |         51 |         7 |   105

data_tabulate(efc, "c172code", by = efc$e16sex, proportions = "row", include_na = FALSE, weights = efc$weights)
#> c172code |       male |     female | Total
#> ---------+------------+------------+------
#> 1        |  5 (62.5%) |  3 (37.5%) |     8
#> 2        | 32 (50.0%) | 32 (50.0%) |    64
#> 3        |  3 (21.4%) | 11 (78.6%) |    14
#> ---------+------------+------------+------
#> Total    |         40 |         46 |    86


data_tabulate(efc, "c172code", by = "e16sex", proportions = "column")
#> c172code |       male |     female |      <NA> | Total
#> ---------+------------+------------+-----------+------
#> 1        |  5 (11.1%) |  2  (4.0%) | 1 (20.0%) |     8
#> 2        | 31 (68.9%) | 33 (66.0%) | 2 (40.0%) |    66
#> 3        |  4  (8.9%) | 11 (22.0%) | 1 (20.0%) |    16
#> <NA>     |  5 (11.1%) |  4  (8.0%) | 1 (20.0%) |    10
#> ---------+------------+------------+-----------+------
#> Total    |         45 |         50 |         5 |   100

data_tabulate(efc, "c172code", by = "e16sex", proportions = "column", include_na = FALSE)
#> c172code |       male |     female | Total
#> ---------+------------+------------+------
#> 1        |  5 (12.5%) |  2  (4.3%) |     7
#> 2        | 31 (77.5%) | 33 (71.7%) |    64
#> 3        |  4 (10.0%) | 11 (23.9%) |    15
#> ---------+------------+------------+------
#> Total    |         40 |         46 |    86

efc |>
  data_group("e42dep") |>
  data_tabulate("c172code", by = "e16sex", proportions = "row")
#> Grouped by e42dep (1)
#> 
#> c172code |       male |       <NA> | Total
#> ---------+------------+------------+------
#> 2        | 2 (100.0%) | 0   (0.0%) |     2
#> <NA>     |   0 (NaN%) |   0 (NaN%) |     0
#> ---------+------------+------------+------
#> Total    |          2 |          0 |     2
#> 
#> Grouped by e42dep (2)
#> 
#> c172code |      male |    female |      <NA> | Total
#> ---------+-----------+-----------+-----------+------
#> 2        | 2 (50.0%) | 2 (50.0%) | 0  (0.0%) |     4
#> <NA>     |  0 (NaN%) |  0 (NaN%) |  0 (NaN%) |     0
#> ---------+-----------+-----------+-----------+------
#> Total    |         2 |         2 |         0 |     4
#> 
#> Grouped by e42dep (3)
#> 
#> c172code |      male |     female |      <NA> | Total
#> ---------+-----------+------------+-----------+------
#> 1        | 2 (50.0%) |  2 (50.0%) | 0  (0.0%) |     4
#> 2        | 4 (25.0%) | 11 (68.8%) | 1  (6.2%) |    16
#> 3        | 1 (16.7%) |  5 (83.3%) | 0  (0.0%) |     6
#> <NA>     | 1 (50.0%) |  0  (0.0%) | 1 (50.0%) |     2
#> ---------+-----------+------------+-----------+------
#> Total    |         8 |         18 |         2 |    28
#> 
#> Grouped by e42dep (4)
#> 
#> c172code |       male |     female |      <NA> | Total
#> ---------+------------+------------+-----------+------
#> 1        |  3 (75.0%) |  0  (0.0%) | 1 (25.0%) |     4
#> 2        | 23 (54.8%) | 18 (42.9%) | 1  (2.4%) |    42
#> 3        |  3 (30.0%) |  6 (60.0%) | 1 (10.0%) |    10
#> <NA>     |  3 (42.9%) |  4 (57.1%) | 0  (0.0%) |     7
#> ---------+------------+------------+-----------+------
#> Total    |         32 |         28 |         3 |    63
#> 
#> Grouped by e42dep (NA)
#> 
#> c172code |       male |     female |       <NA> | Total
#> ---------+------------+------------+------------+------
#> 2        | 0   (0.0%) | 2 (100.0%) | 0   (0.0%) |     2
#> <NA>     | 1 (100.0%) | 0   (0.0%) | 0   (0.0%) |     1
#> ---------+------------+------------+------------+------
#> Total    |          1 |          2 |          0 |     3

# errors
data_tabulate(efc$c172code, by = "e16sex")
#> Error: If `by` is a string indicating a variable name, `x` must be a data
#>   frame.

data_tabulate(efc$c172code, by = efc$e16sex[-1])
#> Error: Length of `by` must be equal to length of `x`.

data_tabulate(efc, "c172code", by = efc$e16sex[-1])
#> Error: Length of `by` must be equal to number of rows in `x`.

data_tabulate(efc, "c172code", by = "c16sex")
#> Error: The variable specified in `by` was not found in `x`. Did you mean
#>   "e16sex"?

data_tabulate(efc, "c172code", by = c("e16sex", "e42dep"))
#> Error: If `by` is a string indicating a variable name, `by` must be of length
#>   1.
#>   You may use `data_group()` to group by multiple variables, then call
#>   `data_tabulate()`.
```

<sup>Created on 2024-02-13 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

### CODE REVIEW

**Comment by strengejacke** [comment]
@etiennebacher WDYT? Enhancing `data_tabulate()`, to be able to print crosstables, using the `by` argument. I think it's better to get crosstables from `data_tabulate()` instead of adding another function?

File: NEWS.md [hunk 1]
```diff
@@ -5,7 +5,10 @@ CHANGES
 * `data_modify()` gets three new arguments, `.at`, `.if` and `.modify`, to modify
   variables at specific positions or based on logical conditions.
 
-* `data_tabulate()` gets a `weights` argument, to compute weighted frequency tables.
+* `data_tabulate()` was revised and gets several new arguments: a `weights`
+  argument, to compute weighted frequency tables. `include_na` allows to include
+  or omit missing values from the table. Furthermore, a `by` argument was added,
+  to compute crosstables.
```
Comment (etiennebacher) [inline NEWS.md:11]:
```suggestion
  to compute crosstables (#479, #481).
```

File: R/data_tabulate.R [hunk 2]
```diff
@@ -1,32 +1,47 @@
...
 #' for printing.
+#' @param include_na Logical, if `TRUE`, missing values are included in the
+#' frequency or crosstable, else missing values are omitted.
 #' @param collapse Logical, if `TRUE` collapses multiple tables into one larger
 #' table for printing. This affects only printing, not the returned object.
 #' @param weights Optional numeric vector of weights. Must be of the same length
 #' as `x`. If `weights` is supplied, weighted frequencies are calculated.
+#' @param proportions Optional character string, indicating the type of
+#' percentages to be calculated. Only applies to crosstables, i.e. when `by` is
+#' not `NULL`. Can be `"row"`, `"column"` or `"cell"`, to calculate row, column
```
Comment (etiennebacher) [inline R/data_tabulate.R:26]: I think `"cell"` is a bit unclear as we actually compute the proportion over the full table and not just a cell. I think `"full"` would be better, WDYT?

File: R/data_tabulate.R [hunk 3]
```diff
@@ -78,25 +125,52 @@ data_tabulate.default <- function(x, drop_levels = FALSE, weights = NULL, name =
     x <- droplevels(x)
   }
 
-  # check for correct length of weights - must be equal to "x"
-  if (!is.null(weights) && length(weights) != length(x)) {
-    insight::format_error("Length of `weights` must be equal to length of `x`.")
+  # validate "weights"
+  weights <- .validate_tableweights(weights, x)
```
Comment (etiennebacher) [inline R/data_tabulate.R:129]: can you rename this to `.validate_table_weights()`?

**Review by etiennebacher — CHANGES_REQUESTED** [review]
Looks great, thanks! I just have some minor comments

File: R/data_tabulate.R (hunk 2 above)
Comment (strengejacke) [reply R/data_tabulate.R:26]: Yes, I agree, "cell" is unclear. We actually have row or column percentages, and relative frequencies ("full"). But something like "rel" or "relfrq" or whatsoever doesn't sound better, and doesn't fit nicely with "row" and "column". Maybe we can go with "full" for now.

**Review by etiennebacher — APPROVED** [review]
Thanks!

**Comment by strengejacke** [comment]
Not sure why the snapshot tests did not render properly, should now work. Once tests pass, I'll merge.

**Pull request merged**